Minimise reliance on specific descriptor argument names (for #836)

TeranIvy opened the pull request:
As outlined in the description for issue #836, this PR introduces the foundations to determine intrinsic (primitive) Fortran type of the LFRic API data structures from the argument descriptor metadata and minimises reliance on the specific `LFRicArgDescriptor` names in declaration classes in the LFRic API.
The PR also consolidates some existing features that are used by both LFRic (Dynamo 0.3) and GOcean 1.0 APIs.

- The `type` property of `LFRicArgDescriptor` and `GO1p0Descriptor` subclasses of the base `Descriptor` class is now moved directly into the base class. There are several `psyGen` functions that extract arguments that are based on descriptor type (for instance, `args_filter` that is also called from Loop class) so it made sense to make `type` one of fundamental base class properties.  This led to some small adjustment in GOcean1.0 API and the relevant tests. More importantly, it highligted that we use a Python keyword (`type`) to name one of the argument properties, which is not a good practice (and makes `pylint` complain). Modifying the property name would probably be a good idea (in a different issue, of course).

- `LFRicArgDescriptor` now has `data_type` property and `_dtype_from_descr` function to determine the Fortran type of argument data from the descriptor if provided. This is currently implemented for scalars only as fields and operators are `real`. Issues #774 and #817 will introduce metadata for data types as well as using them to determine intrinsic types of argument data. They will be the same as the current scalar names whilst there will be other name for scalar arguments. 

- The `intrinsic_type` abstract property for the kernel arguments is added to the base `Argument` class. This property was already present in the `GOKernelGridArgument` class and it is now also implemented in the `DynKernelArgument` class. This was done to enable the data type information to be stored in the kernel argument information. I went for the name `intrinsic_type` because there is already a lot of use of `intrinsic` in PSyIR.

- LFRic API relies heavily on descriptor names as proxies for the argument type. There is already `is_scalar()` method in the base `KernelArgument` class so adding `is_field()` to the `DynKernelArgument` class seemed as a logical expansion. To be consistent with the base class implementation, `is_operator` in `DynKernelArgument` was changed from property into method. Perhaps property is the way to go, but then it would need to be done in the base class and derived API classes as well (separate issue).
The Boolean methods `is_scalar()`,  `is_field()` and  `is_operator()` have now replaced reliance on descriptor name where appropriate in `dynamo0p3.py` and `psyGen.py`.

- The mapping from API scalar types to `psyGen` scalar types for LFRic will be determined in `LFRicArgDescriptor` via the added `MAPPING_SCALARS_LIST`. This will enable support for the current and the new scalar metadata.

A lot of functionality in PSyclone is tied to the concept of either implicitly assuming data types (e.g. fields are `real`) or using specific descriptor name to infer both argument type and data type (e.g. `GH_REAL` for `real` scalar). Some examples are filtering of PSy layer arguments (`args_filter`) and identifying arguments for decalarations, as well as scalar reduction names (`MAPPING_SCALARS`).  This makes introducing new functionality that does not rely on these assumptions difficult, but refactoring this is beyond the scope of this PR.

Hopefully what is done here will make it easier to support both existing and new LFRic metadata.

Ready for review from either @arporter or @rupertford.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3508,6 +3508,18 @@ def type(self):
 
+    @property
+    @abc.abstractmethod
+    def intrinsic_type(self):
```
The addition of this property highlights that the name of the immediately preceding `type` method is not helpful :-) Again, I think something like `object_type` or perhaps we should sub-class `Argument` as `ScalarArgument`, `FieldArgument`, `OperatorArgument` etc. etc. I suspect that's the right way to go as we're asking the question "what type is this Argument object?" and the answer is naturally e.g. "a FieldArgument". That's clearly not something for this PR though so would you mind creating an Issue where we can discuss sub-classing `Argument`? Thanks :-)

TeranIvy replied:
How about the above suggestion of `argument_type` in this PR and then subclassing in the later ones?

TeranIvy replied:
Renamed `type` to `argument_type` as discussed above. Created issue #859 to discuss sub-classing `Argument`.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2896,29 +2900,30 @@ def __init__(self, node):
+                        if arg.intrinsic_type == "real":
+                            self._real_scalar_names[intent].append(declname)
+                        if arg.intrinsic_type == "integer":
+                            self._int_scalar_names[intent].append(declname)
```
As this stands, we need a final `elif` to catch cases where the `intrinsic_type` is something we don't support.

TeranIvy replied:
Not any more, as noted in one of the above comments. The invalid intrinsic type is now caught when setting up `DynKernelArgument.intrinsic_type` property that creates intrinsic type from `dynamo0p3.MAPPING_DATA_TYPES` and checks there if the mapping is invalid.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -5462,16 +5475,37 @@ def test_unsupported_halo_read_access():
 
+def test_dynglobalsum_unsupported_argument():
+    ''' Check that an instance of the DynGlobalSum class raises an
+    exception if an unsupported argument type. '''
```
"if an unsupported type....??"

TeranIvy replied:
Oops. Should be "for an unsupported argument type". Changed.

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -415,22 +416,25 @@ def get_stencil(metadata, valid_types):
 
 class Descriptor(object):
-    '''A description of how a kernel argument is accessed
+    ''' A description of how a kernel argument is accessed.
```
"...accessed, constructed from the kernel metadata."

TeranIvy replied:
Thanks for the suggestion, done.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -142,19 +142,21 @@ def args_filter(arg_list, arg_types=None, arg_accesses=None, arg_meshes=None,
-    :param arg_types: List of argument types (e.g. "GH_FIELD")
+    :param arg_list: list of kernel arguments to filter.
+    :type arg_list: list of :py:class:`psyclone.parse.kernel.Descriptor`
+    :param arg_types: list of argument types (e.g. "GH_FIELD").
```
If we moved to sub-classing `Argument` (see discussion elsewhere), this check would just become `isinstance`.

TeranIvy replied:
Agreed. Created #859 to discuss sub-classing `Argument`.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2896,29 +2900,30 @@ def __init__(self, node):
         # Create lists of real and integer scalar arguments
         self._real_scalar_names = {}
         self._int_scalar_names = {}
+        for intent in FORTRAN_INTENT_NAMES:
```
This implementation now contains a mix of the old (assumes just 'real' and 'int') and the new (allows for all VALID_SCALAR_NAMES). Could we instead just use the dict of scalar arguments throughout the class, filtering as required? Alternatively, we could have a new `_scalar_names` dict with each entry indexed by intent and itself consisting of a dict indexed by intrinsic type?
(Presumably you're going to change VALID_SCALAR_NAMES to just be "gh_scalar" at some point (future PR) with the type information specified separately?)

TeranIvy replied:
Yes, I thought about that and I think there are two potential solutions:
1. Modify  `unique_declns_by_intent` and `unique_declarations` to accept a list of names instead of just one name (this would cover as for the change in number of argument names) and do filterinb by intrinsic type in the API;
2. 1 + modify these two functions to also do filtering by intrinsic type .

I am happy to do either, as they are only used in the LFRic API for now.

TeranIvy replied:
I went for the first solution, and that is to modify `unique_declns_by_intent` and `unique_declarations` to accept a list of names instead of just one name (also did it for `unique_proxy_declarations` for consistency).

The `DynScalarArgs.__init__` method is now rewritten to first return a dictionary of scalar arguments by intent for both Invoke and Kernel objects. Then the arguments are filtered by intrinsic types to populate the dictionaries for `real` and `integer` scalars.

As noted from a comment above, the invalid intrinsic type is now caught when setting up `DynKernelArgument.intrinsic_type` property (creates intrinsic type from `dynamo0p3.MAPPING_DATA_TYPES` and checks the mapping there).

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9020,31 +9052,32 @@ def scalars(self):
         self._kernel_args = kernel_args
         self._vector_size = arg_meta_data.vector_size
         self._type = arg_meta_data.type
+        self._intrinsic_type = arg_meta_data.data_type
```
I suspect (but am not sure) that this is the point at which we should map from metadata information into DSL information. i.e. what the intrinsic type of the object is. I think the type should be one of `psyir.symbols.ScalarType.Intrinsic.INTEGER/REAL/...` and should be stored in some map (dict). Ultimately this could come from the configuration file but obviously not yet :-)
At the moment, the mapping from metadata to intrinsic type is done in LFRicArgDescriptor which is not a bad place but my feeling is that that class should deal purely with the kernel metadata. (Possibly it and those related to it should be renamed to e.g. LFRicArgMetadata to make this clear.) The current class brings together the information from the Algorithm and the kernel metadata and is therefore maybe a more natural place to introduce this type information.

TeranIvy replied:
Yes, I thought about it, too. My initial implementation in #774 was to just leave `LFRicArgDescriptor.data_type` as metadata. The mapping from descriptor to intrinsic type would happen in the `DynKernelArgument` class.
For instance, descriptor `data_type` for e.g. `(GH_FIELD, GH_READ, W2)` would be `GH_REAL` and `(GH_SCALAR, GH_INTEGER, GH_READ)` would be `GH_INTEGER`. Argument `intrinsic_type` would then be determined by some mapping from that.

The solution implemented here was to a) put something in place until we have metadata for indicating the intrinsic type and b) limit the simultaneous support for the current and the new metadata just to the `LFRicArgDescriptor` class. For instance, the support for fields would be something like
```
if current metadata (e.g. (GH_FIELD, GH_READ, W2)) then
    data_type = `real`
elif new metadata (e.g. (GH_FIELD, GH_REAL, GH_READ, W2)) then
    data_type = descriptor_from_metadata(arg_type.args[1])
```
and for scalars
```
if current metadata (e.g. (GH_INTEGER, GH_READ)) then
    data_type = descriptor_from_metadata(arg_type.args[0])
elif new metadata (e.g. (GH_SCALAR, GH_INTEGER, GH_READ)) then
    data_type = descriptor_from_metadata(arg_type.args[1])
```

A  way out of this conundrum would be to simply define `data_type` as `GH_REAL` or `GH_INTEGER` in the descriptor class in this PR, i.e.
```
if current metadata (e.g. (GH_FIELD, GH_READ, W2)) then
    data_type = GH_REAL
elif new metadata (e.g. (GH_FIELD, GH_REAL, GH_READ, W2)) then
    data_type = descriptor_from_metadata(arg_type.args[1])
```
and
```
if current metadata (e.g. (GH_INTEGER, GH_READ)) then
    data_type = arg_type.args[0]
elif new metadata (e.g. (GH_SCALAR, GH_INTEGER, GH_READ)) then
    data_type = arg_type.args[1]
```
and introduce the mapping to intrinsic type somewhere else (as said above, my initial idea was `DynKernelArgument` class but I am open to suggestions). What do you think? 
Of course, that would mean essentially duplicating the current scalar names but we would do that anyway if we want to support both

TeranIvy replied:
I decided to implement the second solution where the `LFRicArgDescriptor.data_type` is one of `LFRicArgDescriptor.VALID_ARG_DATA_TYPES = ["gh_real", "gh_integer"]`. The scalar `data_type` is now simply its name but that will change in #774. Also, #774 will establish a check whether `data_type` metadata descriptor is one of `LFRicArgDescriptor.VALID_ARG_DATA_TYPES`. Fields and operators now have `gh_real` but that will be extended in #817. 

`DynKernelArgument.intrinsic_type` is now set via a `dynamo0p3.MAPPING_DATA_TYPES` mapping with the `InternalError` for a wrong key.

sergisiso commented:
>  As I said in the telco, I think we should move away from having type as "real" or "integer" and use psyir.symbols.ScalarType.Intrinsic.XXX. My only concern there is that we're using something from within a class with "Scalar" in the title but that's how @rupertford and @sergisiso have named it.

We discussed how to compare types with @rupertford on the the PR that introduced the different types but decided not to solve type comparisons yet as we didn't see clearly the use cases. My hope was/is to overload the `Type.__eq__` method in all types so that we only have to declare once the valid types e.g. `LFRIC_REAL = ScalarType(ScalarType.Intrinsic.REAL, precision_datasymbol)` and then we can do:
```
assert sym.type == LFRIC_REAL
assert sym.type != LFRIC_INTEGER
assert sym.type != ArrayType(LFRIC_TYPE, [2,3])
```
With this we can compare without knowing the internal Types properties (intrinsic, precision ...) or even completely different types (Arrays, Structures, ..) in a single statement.

But of course this is not available in the code yet.

sergisiso commented:
The relevant discussion is on #744

sergisiso commented:
Reading again this PR description I realize that this compares something outside PSyIR types (metadata-descriptions) so what I said may not be applicable here at the moment but still I hope this clarifies the idea we had for PSyIR Types comparisons.

rupertford commented:
> As I said in the telco, I think we should move away from having type as "real" or "integer" and use `psyir.symbols.ScalarType.Intrinsic.XXX`. My only concern there is that we're using something from within a class with "Scalar" in the title but that's how @rupertford and @sergisiso have named it. Hopefully this usage is in line with their thinking?

real and integer are scalars :-)

TeranIvy commented on src/psyclone/parse/kernel.py:
```diff
@@ -468,7 +472,22 @@ def mesh(self):
         return self._mesh
 
+    @property
+    def type(self):
```
I would rather go for  `arg_type` or `argument_type` as it maps nicely into the names used in various `psyGen` functions. This, of course, means that it would be god to change `type` to something else not only in the `Descriptor` and the children classes but also in `Argument` and the related children classes. It is a lot of code changes as `descriptor.type` and `argument.type` are heavily used!

TeranIvy replied:
Changed to `argument_type` as this name is used a lot in `psyGen` for applications of this property.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9110,62 +9147,115 @@ def ref_name(self, function_space=None):
+                "DynKernelArgument.ref_name(fs): Found unsupported argument "
+                "type '{0}'.".format(self._type))
+
+    def is_scalar(self):
```
I agree with you but I think it would then be good to do this from the base `KernelArgument` where `is_scalar()` is first defined. This means changes there and in `GOcean` classes, code and tests as well. I'm happy to consolidate them in this PR if you're happy with it.

TeranIvy replied:
P.S. As said in a comment above, I changed all these methods into properties (starting with the abstract property `is_scalar` in the base `KernelArgument` class).

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9110,62 +9147,115 @@ def ref_name(self, function_space=None):
+        if self.is_field():
             return "vspace"
-        if self.is_operator:
+        if self.is_operator():
```
For consistency with `is_scalar()` which is not a property but a method because it is set that way in the base `KernelArgument` class. As I said in a comment below, I am happy to do method to property from ground up to the APIs that use it.

TeranIvy replied:
P.S. There were not so many instances in the code so I changed the method into a property (starting with the abstract property in the `KernelArgument` class).

arporter replied:
That's great - thanks Iva. I think it's much better now.

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -66,8 +66,8 @@
 MAPPING_SCALARS_LIST = ["iscalar", "rscalar"]
-# Mapping from domain-specific scalar-type names to internal scalar
-# types. Can be overridden in domain-specific modules.
+# Mapping from domain-specific scalar-type names to internal scalar types
```
Well, from the code that the list is used for it seems they are used for identifying arguments that can be valid reductions. I am not particularly happy with this mode of identification as I think the checks should really be based on `is_scalar` and then leave it to a specific API to define what scalars are allowed to be reductions (e.g. only `real` in LFRic API). I stopped short of rewriting the related code as I thought this PR would grow out of proportions.

TeranIvy replied:
Changed back to the original description.

TeranIvy commented:
> 
> We discussed how to compare types with @rupertford on the the PR that introduced the different types but decided not to solve type comparisons yet as we didn't see clearly the use cases. My hope was/is to overload the `Type.__eq__` method in all types so that we only have to declare once the valid types e.g. `LFRIC_REAL = ScalarType(ScalarType.Intrinsic.REAL, precision_datasymbol)` and then we can do:
> 
> ```
> assert sym.type == LFRIC_REAL
> assert sym.type != LFRIC_INTEGER
> assert sym.type != ArrayType(LFRIC_REAL, [2,3])
> ```
> 
> With this we can compare without knowing the internal Types properties (intrinsic, precision ...) or even completely different types (Arrays, Structures, ..) in a single statement.
> 
> But of course this is not available in the code yet.

The above example assumes that `LFRIC_REAL` and `LFRIC_INTEGER` are scalars - but we will have other argument types that have `real` and `integer` data types, fields for instance. This was my main motivation to separate `argument_type` and `data_type` (e.g. `(GH_SCALAR, GH_REAL, GH_READ)` and `(GH_FIELD, GH_INTEGER, GH_WRITE, W3)`).

rupertford commented:
> @sergisiso , what worries me about the above example is that it assumes `LFRIC_REAL` and `LFRIC_INTEGER` are scalars - but we will have other argument types that have `real` and `integer` data types, fields for instance. This was my main motivation to have separate descriptors for `argument_type` and `data_type` (e.g. `(GH_SCALAR, GH_REAL, GH_READ)` and `(GH_FIELD, GH_INTEGER, GH_WRITE, W3)`).

@TeranIvy, that is the point I was trying to make at the telco and in my comment above. The datatypes in the PSyIR are scalar datatypes and array datatypes. They represent the declarations of data. So, I would rephrase what you say above. The example does not assume LFRIC_REAL and LFRIC_INTEGER are scalars, they *are* scalars by definition and design. This is why I don't think that the concepts are the same. You could re-use the names but that would be confusing imho. I think there should be a separate enumerated type.

TeranIvy commented:
> 
> @TeranIvy, that is the point I was trying to make at the telco and in my comment above. The datatypes in the PSyIR are scalar datatypes and array datatypes. They represent the declarations of data. So, I would rephrase what you say above. The example does not assume LFRIC_REAL and LFRIC_INTEGER are scalars, they _are_ scalars by definition and design. This is why I don't think that the concepts are the same. You could re-use the names but that would be confusing imho. I think there should be a separate enumerated type.

Aaah, I am with you now @rupertford and @sergisiso. If I understood correctly what you both said, sometimes there could be 1-2-1 mapping between the current `psyGen` `Argument` types and PSyIR symbols (as in the scalar case) and sometimes not. For instance, LFRic fields would be something like a Structure as they hold more information (e.g. `dofmap`, `undf`) than a simple array of data values. NEMO fields on the other hand seem to me as Arrays. 
Did I understand this correctly? If I did, I don't feel comfortable trying to join two concepts in this PR until we have more clarity on how the translation of concepts will work (I do think, however, that it would be good to think about it sooner rather than later). I am happy to stick to the `intrinsic_type` that uses a string for declarations in this PR if @arporter is happy with that, too.

TeranIvy commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -292,21 +300,25 @@ def _validate_field(self, arg_type):
-                       len(arg_type.args), arg_type))
+                       self._nargs, arg_type))
+
+        # Field data_type is "real" for now, but will be determined by
```
I went for `_validate` originally as there seems to be many `_validate_*` methods in LFRic API that do the mix of checking and property setting. I am happy to rename the methods but it would be good to have some developer guidance when to use `_validate*` and when to use `_init*` .

TeranIvy replied:
Renamed `_validate_*` to `_init_*` methods as requested.

arporter replied:
Thanks.

TeranIvy commented on src/psyclone/parse/kernel.py:
```diff
@@ -415,22 +416,25 @@ def get_stencil(metadata, valid_types):
 
     '''
-    def __init__(self, access, space, stencil=None, mesh=None):
+    def __init__(self, access, space, stencil=None, mesh=None, type=None):
```
Changed to `argument_type` as this name is used a lot in `psyGen` applications of this property.

arporter replied:
Great.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2805,6 +2807,8 @@ class DynCellIterators(DynCollection):
                           :py:class:`psyclone.dynamo0p3.DynInvoke`
+
+    : raises GenerationError: if an Invoke has no field or operator arguments.
     '''
```
Done.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2886,8 +2890,8 @@ class DynScalarArgs(DynCollection):
 
-    :raises InternalError: if an unrecognised type of scalar argument is \
-                           encountered.
+    :raises InternalError: if an unrecognised intrinsic type of the scalar \
```
Done.

TeranIvy replied:
This internal error turned out to be obsolete as the invalid intrinsic type is now caught when setting up `DynKernelArgument.intrinsic_type` property. This property creates intrinsic type from `dynamo0p3.MAPPING_DATA_TYPES` so there is an `InternalError` there when the mapping is invalid.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9020,31 +9052,32 @@ def scalars(self):
-    arguments as specified by the kernel argument metadata. '''
+    '''
+    This class provides information about individual LFRic kernel call
+    arguments as specified by the kernel argument metadata.
```
Thanks for the suggestion, done!

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9110,62 +9147,115 @@ def ref_name(self, function_space=None):
-                    "function_space_to '{3}'".format(
+                    "DynKernelArgument.ref_name(fs): Function space '{0}' "
+                    "is one of the 'gh_operator' function spaces '{1}' but "
+                    "is not being returned by either function_space from "
```
Done.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9110,62 +9147,115 @@ def ref_name(self, function_space=None):
     def type(self):
-        ''' Returns the type of this argument. '''
+        '''
+        :returns: the API type of this argument.
```
Done.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9110,62 +9147,115 @@ def ref_name(self, function_space=None):
+        :returns: True if this kernel argument represents a scalar, \
+                  False otherwise.
+        :rtype: bool
+
```
Deleted.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9110,62 +9147,115 @@ def ref_name(self, function_space=None):
         Getter for the mesh associated with this argument
-        :return: Mesh associated with argument (GH_FINE or GH_COARSE)
+
+        :returns: mesh associated with argument (GH_FINE or GH_COARSE).
```
Removed.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9110,62 +9147,115 @@ def ref_name(self, function_space=None):
         '''
         :returns: the proxy name for this argument.
         :rtype: str
+
```
As said above, deleted.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9225,27 +9326,46 @@ def function_space(self):
+        '''
+        Returns the expected finite element function space for a kernel
+        argument as specified by the kernel argument metadata: a single
+        function space for a field and function_space_from for an operator.
```
Rephrased.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9225,27 +9326,46 @@ def function_space(self):
+        function space for a field and function_space_from for an operator.
+
+        :returns: function space for this argument.
+        :rtype: :py:class:`psyclone.dynamo0p3.FunctionSpace`
```
Done.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9279,39 +9401,45 @@ def intent(self):
-        '''Return stencil information about this kernel argument if it
-        exists. The information is returned as a DynStencil object.'''
+        '''
+        Returns stencil information about this kernel argument if it
```
Deleted the redundant description and slightly rephrased `:returns:`.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9279,39 +9401,45 @@ def intent(self):
-        :return: True if this kernel argument represents an operator,
-                 False otherwise.
-        :rtype: bool
+        Sets stencil information for this kernel argument. The information
```
Done.

TeranIvy commented on src/psyclone/dynamo0p3_builtins.py:
```diff
@@ -45,14 +45,16 @@
 BUILTIN_DEFINITIONS_FILE = "dynamo0p3_builtins_mod.f90"
 
 # The types of argument that are valid for built-in kernels in the
 # Dynamo 0.3 API
```
Done.

TeranIvy commented on src/psyclone/dynamo0p3_builtins.py:
```diff
@@ -161,9 +163,9 @@ def _validate(self):
             if arg.type not in VALID_BUILTIN_ARG_TYPES:
                 raise ParseError(
                     "In the Dynamo 0.3 API an argument to a built-in kernel "
```
Done.

TeranIvy commented on src/psyclone/parse/kernel.py:
```diff
@@ -468,7 +472,22 @@ def mesh(self):
+        return self._type
+
     def __repr__(self):
+        '''
```
Thanks, deleted.

TeranIvy commented on src/psyclone/gocean1p0.py:
```diff
@@ -1985,29 +1994,27 @@ def __init__(self, kernel_name, kernel_arg):
+                             type=self._type)
 
     def __str__(self):
+        '''
```
Deleted.

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -2403,28 +2406,25 @@ def zero_reduction_variable(self, parent, position=None):
+                "found '{0}'.".format(var_arg.type))
+        # Generate the reduction variable
+        var_data_type = var_arg.intrinsic_type
+        if var_data_type == "real":
```
I think we agreed it's not yet practical so I left it as it is.

TeranIvy commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -524,25 +545,39 @@ def _validate_scalar(self, arg_type):
+        :rtype: str
+
+        '''
+        dtype = str(dtype_descriptor).lower()
```
I think we agreed that we're not using the PSyIR types just yet so I added a mapping from meta-data name to datatype in the LFRicArgDescriptor class.

arporter replied:
I'm happyish with this for now. Maybe we need a separate enumeration of primitive types for meta-data purposes but that's something for later.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -134,10 +134,10 @@
 # ---------- psyGen mappings ------------------------------------------------ #
-# Mappings used by non-API-Specific code in psyGen
-# psyGen ["iscalar", "rscalar"] translate to LFRic ["gh_integer", "gh_real"]
+# Mappings used for scalar reductions by non-API-Specific code in psyGen.
```
Rephrased.

TeranIvy commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -101,12 +101,17 @@ class LFRicArgDescriptor(Descriptor):
     # perform prolongation/restriction).
     VALID_MESH_TYPES = ["gh_coarse", "gh_fine"]
+
+    # LFRic names for valid scalar reductions
```
We can use `LFRicArgDescriptor.VALID_SCALAR_NAMES` for `psyGen.MAPPING_SCALARS_LIST` for now, but in #774 I will need to refactor that mapping. I have removed it for now.

TeranIvy commented:
Addressed review comments and updated the branch with the latest master. Ready for another review from @arporter.

arporter commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -316,9 +317,9 @@ def _validate_field(self, arg_type):
+        # Field data_type is "gh_real" for now, but will be determined by
         # metadata descriptor as the second argument in issue #817
-        self._data_type = "real"
+        self._data_type = LFRicArgDescriptor.VALID_ARG_DATA_TYPES[0]
```
I think you need to stick with "gh_real" here rather than just arbitrarily index into the list of valid arg types. This is OK as it's a stepping stone towards #817 as you already say in the comment.

TeranIvy replied:
Done.

arporter commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -455,9 +456,9 @@ def _validate_operator(self, arg_type):
+        # Operator data_type is "gh_real" for now, but will be determined by
         # metadata descriptor as the second argument in issue #817
-        self._data_type = "real"
+        self._data_type = LFRicArgDescriptor.VALID_ARG_DATA_TYPES[0]
```
Again, please just use "gh_real" here.

TeranIvy replied:
Done.

arporter commented on src/psyclone/domain/lfric/lfric_arg_descriptor.py:
```diff
@@ -512,9 +513,9 @@ def _validate_scalar(self, arg_type):
         '''
         # Check whether something other than a scalar is passed in
-        if self._type not in LFRicArgDescriptor.VALID_SCALAR_NAMES:
+        if self._argument_type not in LFRicArgDescriptor.VALID_SCALAR_NAMES:
```
Just a note to say this doesn't look right (although it's not the fault of this PR) - we shouldn't be comparing the 'type' of something with a list of 'valid scalar names'. I'd expect 'valid scalar types'. I'm not asking you to fix this - just making a note of it.

TeranIvy replied:
I agree, it is clunky. As you said, perhaps as a part of another PR - it would be good to think about how the present `psyGen` classes/functions' names relate to names of similar structures in PSyIR.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2904,40 +2913,34 @@ class DynScalarArgs(DynCollection):
+                if arg.intrinsic_type == "real":
+                    self._real_scalar_names[intent].append(declname)
+                if arg.intrinsic_type == "integer":
+                    self._int_scalar_names[intent].append(declname)
```
For defensive purposes this should be:

    if real:
        ...
    elif integer:
        ...
    else:
        raise InternalError(...)

(That way, if/when we should extend the list of supported scalar types we'll get a nice error message at this point.)

TeranIvy replied:
Done. Restored and adapted the `test_dynscalars_err` to `test_dynscalars_stub_err` in `dynamo0p3_stubgen_test.py` for testing the kernel stub. Added `test_dynscalars_call_err` in `dynamo0p3_test.py` to test the kernel call.

arporter replied:
Please include the valid types in the error message.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -95,6 +95,10 @@
 # ---------- Fortran datatypes ---------------------------------------------- #
 SUPPORTED_FORTRAN_DATATYPES = ["real", "integer", "logical"]
```
It's about here I start getting uneasy. This list of types is used to map from meta-data types to Fortran types, at which point we should be using PSyIR types? However, this would only work properly (as pointed out by @rupertford) if the PSyIR defined primitive types and currently it does not. Not something for this PR though.

TeranIvy replied:
This list of Fortran types is also used to set and check the mapping from Fortran type to default precision for the generated LFRic PSy-layer code in `configuration.py`. We will be able to modify some of it with the mixed pecision work. In the absence of the PSyIR functionality this was the closest to a consistent approach I could think of, as using the `SUPPORTED_FORTRAN_DATATYPES` list for `MAPPING_DATA_TYPES` acts as a bridge between the kernel descriptors for Fortran data type and the default precisions these types have. It is clunky, I agree, but I also think that it at least provides useful information about the type and precision properties for PSyIR Symbols when we start using them for kernel/algorithm arguments.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -136,10 +140,10 @@
 
 # ---------- psyGen mappings ------------------------------------------------ #
-# Mappings used for scalar reductions by non-API-Specific code in psyGen.
+# Mappings used by non-API-Specific code in psyGen.
```
This will ultimately be 'proper' PSyIR but that isn't going to happen for a while yet.

TeranIvy replied:
I suspect I will have to find a way around it in #774, as I will have just one scalar name.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -9151,17 +8029,27 @@ def intrinsic_type(self):
         '''
+
+        try:
+            self._intrinsic_type = MAPPING_DATA_TYPES[
```
I don't think we should be setting `_intrinsic_type` here. Could we do it in the constructor or just do away with it altogether?

TeranIvy replied:
Moved to the `DynKernelArgument` constructor together with the `InternalError` check and description. Adjusted the `test_arg_intrinsic_type_error` in `dynamo0p3_test.py` accordingly.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -566,11 +566,12 @@ def unique_declarations(self, argument_type, access=None):
-                "Invoke.unique_declarations() called with an invalid argument "
-                "type. Expected one of {0} but found '{1}'".
-                format(str(VALID_ARG_TYPE_NAMES), argument_type))
+        for argtype in argument_types:
```
any works on iterators so we shouldn't need the list? i.e. `if any(argtype not in VALID_ARG_TYPE_NAMES for argtype in argument_types):`

TeranIvy replied:
Very elegant but unfortunately it only returns `True` or `False` but not the type of the offending argument nor how to find it, and that is useful for the error message.
Then I tried this:
```
        invalid_arg = [arg for arg in argument_types if arg not in VALID_ARG_TYPE_NAMES]
        if invalid_arg:
            raise InternalError(...
```
and that returns a list. I figured out this is actually correct behaviour and also practical. Now that we have a list of argument types instead of one there is a possibility that more than one will be invalid and it is better to list them all than check for one at a time. I modified the error messages and docstrings accordingly.

TeranIvy replied:
@rupertford, what you wrote works but also does not return the names of invalid argument types for a useful error message. I decided to go for the list comprehension that returns the names of invalid types.

arporter replied:
I would have just printed the whole list in the event of an error but it matters not :-)

TeranIvy replied:
Ah, I am with you now! Went for Rupert's comprehension and just printed the whole list.

TeranIvy commented on src/psyclone/tests/dynamo0p3_stubgen_test.py:
```diff
@@ -61,46 +62,6 @@ def setup():
     Config.get().api = "dynamo0.3"
 
 
-def test_kernel_stub_invalid_scalar_argument():
```
`test_kernel_stub_invalid_scalar_argument` was moved to `arg_ordering_test.py` in PR #825. As said in a comment above, `test_dynscalars_stub_err` is restored and adapted.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -252,21 +253,21 @@ def test_ad_int_scalar_type_no_sum():
         _ = DynKernMetadata(ast, name=name)
     assert ("reduction access 'gh_sum' is only valid with a real scalar "
-            "argument, but scalar 'gh_integer' with 'integer' data type "
+            "argument, but scalar 'gh_integer' with 'gh_integer' data type "
```
Yes, that is the idea, to have `gh_scalar` as argument name/type and `gh_integer` or `gh_real` as data type. I do agree it looks odd, however it is a stepping stone to #774.

TeranIvy commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -2945,13 +2973,13 @@ def test_arg_descriptor_real_scalar():
     # Check LFRicArgDescriptor argument properties
-    assert scalar_descriptor.type == "gh_real"
-    assert scalar_descriptor.data_type == "real"
+    assert scalar_descriptor.argument_type == "gh_real"
```
Yep, in #774. My last two months-worth of pull requests starting with the restructuring of `LFRicArgDescriptor` are basically chunks of work that came out of #774 (which means I'll probably have to start with a fresh branch there).

TeranIvy commented on src/psyclone/parse/kernel.py:
```diff
@@ -415,22 +416,28 @@ def get_stencil(metadata, valid_types):
+                      on.
+    :param dict stencil: type of stencil access for this argument. \
+                         Defaults to None if the argument is not supplied.
+    :param string mesh: which mesh this argument is on. Defaults to None \
```
Done.

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -620,19 +621,20 @@ def unique_declns_by_intent(self, argument_type):
-                "Invoke.unique_declns_by_intent() called with an invalid "
-                "argument type. Expected one of {0} but found '{1}'".
-                format(str(VALID_ARG_TYPE_NAMES), argument_type))
+        for argtype in argument_types:
```
As said above, I adapted the comprehension to return a list of invalid argument types. I also modified `unique_proxy_declarations` in LFRic API for consistency.

TeranIvy commented:
Addressed review comments and refreshed the branch. Back to @arporter.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -7816,31 +7862,36 @@ def scalars(self):
         self._vector_size = arg_meta_data.vector_size
-        self._type = arg_meta_data.type
+        self._argument_type = arg_meta_data.argument_type
+        self._intrinsic_type = None
```
I don't think we need this line.

TeranIvy replied:
Removed.

TeranIvy commented:
Changes for the review:
- Removed redundant setting of `intrinsic_type` to `None` in `DynKernelArgument` constructor.
- Added the list of supported intrinsic types into the `DynScalarArgs` constructor `InternalError` as requested and updated the related tests. Also noticed that the error wasn't documented so I added the docstring.
- Changed checks for the invalid argument types in `unique_declarations`, `unique_declns_by_intent` and `unique_proxy_declarations` to use the suggested comprehension `if any(argtype not in VALID_ARG_TYPE_NAMES for argtype in argument_types):` and print the full list of `argument_types` in the error message.

Refreshed the branch with the latest master

TeranIvy commented:
Travis is happy so back to @arporter.
